src/armature/Generator.ts

```diff
@@ -93,52 +189,52 @@ export class Generator {
         return this;
     }
 
-    /**
-     * Tells the generator that more components can be generated somewhere.
-     *
-     * @param {SpawnPoint} spawnPoint The name of the component to spawn and the point at which to
-     * spawn it.
-     */
-    public addDetail(spawnPoint: SpawnPoint) {
-        this.nextSpawnPoints.push(spawnPoint);
+    public generate(params: { start: string; depth?: number }): Model {
+        const instance = new GeneratorInstance(this, () => 0);
+        instance.generate(params);
+
+        return instance.getModel();
     }
 
-    /**
-     * Randomly generate an armature from the current component definitions.
-     *
-     * @param {string} start The name of the component to use as a base.
-     * @param {number?} depth How many iterations of generation should be used.
-     * @returns {Node} The root node of the generated armature.
-     */
-    public generate(params: { start: string; depth?: number }): Node {
-        const { start, depth = 10 } = params;
+    public generateSOMC(params: {
+        start: string;
+        depth?: number;
+        samples?: number;
+        costFn: CostFn;
+    }): Model {
+        const { start, depth = 10, samples = 50, costFn } = params;
+        let instances = range(samples).map(() => new GeneratorInstance(this, costFn));
 
-        // Clear spawn points
-        this.nextSpawnPoints.length = 0;
+        // Seed instances with starting state
+        instances.forEach((instance: GeneratorInstance) => instance.initialize(start));
 
-        // Create root node and initial spawn point
-        const root = new Node();
-        root.createPoint('spawn', { x: 0, y: 0, z: 0 });
-        this.addDetail({ component: start, at: root.point('spawn') });
+        range(depth).forEach((iteration: number) => {
+            // Step 1: grow samples
+            instances.forEach((instance: GeneratorInstance) => instance.advance());
 
-        // Run `depth` rounds of generation
-        range(depth).forEach(() => {
-            this.cycleSpawnPoints();
-            this.spawnPoints.forEach((spawnPoint: SpawnPoint) => {
-                const generator = this.getGenerator(spawnPoint.component);
-                generator(spawnPoint.at);
-            });
-        });
+            // Step 2: if there will be more iterations, do a weighted resample
+            if (iteration + 1 !== depth) {
+                const total = instances.reduce((accum: number, instance: GeneratorInstance) => {
+                    return accum + 1 / instance.getCost();
+                }, 0);
+                instances = instances.map(() => {
+                    let sample = this.random() * total;
+                    let picked: GeneratorInstance | null = null;
+                    let i = 0;
+                    do {
+                        picked = instances[i];
+                        sample -= 1 / picked.getCost();
+                        i += 1;
+                    } while (sample > 0);
```

Comment:
Do we have to use a `do`/`while` here? Is there a case where `sample` will be negative in the first loop iteration?

Reply:
It can be rearranged to be a normal `while`, I'll do that